### ISSUE TITLE
Contradictory documentation about pipelines

### ISSUE THREAD

**Issue by lodo1995** [Grokzen/redis-py-cluster#317]
The [Limitations and differences](https://redis-py-cluster.readthedocs.io/en/master/limitations-and-differences.html) page of the documentation states:

> Pipelines do not work the same way in a cluster. In StrictRedis it batch all commands so that they can be executed at the same time when requested. But with RedisCluster pipelines will send the command directly to the server when it is called, but it will still store the result internally and return the same data from .execute(). This is done so that the code still behaves like a pipeline and no code will break. A better solution will be implemented in the future.

The [Pipelines](https://redis-py-cluster.readthedocs.io/en/master/pipelines.html) page, on the other hand, states:

> Just like in redis-py, redis-py-cluster queues up all the commands inside the client until execute is called. But, once execute is called, redis-py-cluster internals work slightly differently. It still packs the commands to efficiently transmit multiple commands across the network. But since different keys may be mapped to different nodes, redis-py-cluster must first map each key to the expected node. It then packs all the commands destined for each node in the cluster into its own packed sequence of commands. It uses the redis-py library to communicate with each node in the cluster.

So, are pipelines batching commands in N batches, where N is the number of partitions, as the second doc suggests, or are pipelines executing all commands one by one, as the first doc suggests?

My feeling is that the first doc might be outdated. In any case, it's clear that at least one of them must be redacted.

**Comment by Grokzen** [comment]
Well this is very technical, but in a way they are both correct still. Even if the second one is more correct then the first one. And the second one is about 2 years older when looking at the changelog.

**Comment by lodo1995** [comment]
Ok. I definitely trust your expertise on this.

But those two pieces of information left me confused, so I have to ask: if I have N partitions in the cluster, and I build a pipeline containing M commands, will the pipeline perform N requests (one per partition) or M (one per command)? Because in my use case I absolutely need efficient pipelining.

**Comment by Grokzen** [comment]
@lodo1995 You can't say for sure really. It do not depend on the size of the cluster or the number of commands. The correct answer is that it depends on how many different slots your commands want to operate on and how distributed those slots are. I dont know what big O notation it would come down to but here a few examples.

If you only have 1 command (a single GET for example) then it would only partition the pipeline into one connection and one request to the node that owns the specific slot.

If you have 3 commands in your pipeline, then the best case scenario is that they all belong to the same slot and in that case then all 3 commands would be packed into the same request for one node. The worst case scenario in this case is that you try to operate on 3 slots that is located on 3 different nodes. Then the pipeline will split the commands into 3 different stacks and send each stack to the correct node, wait for the result back in blocking fashion and then reasemble the pipeline into the correct order that you defined in your client code.

So with this in mind, you can try to optimize the problem as many would argue that in a cluster environment you have to more carefully design your key patterns to make as many keys and operations as possible work with a single node. So instead of having SET/GET commands for attributes on an virtual object, you would instead move to a hash so you can have multiple attributes, but only work with one slot. Same with crossslot functionality that you carefully design or try to scramble/create random keys that you can ensure is on the same node/slot so you avoid crossslot operations for example. Or you have to more carefully manually design your keys so if you do some operation and caluclations with sets or some of the other more complicated types, they will till be doable on the same node to avoid crossslot error or not supported errors from the client side.

**Comment by Grokzen** [comment]
Pipelines is super difficult really as you also have to consider a ton of wierd corner cases and race conditions and order of execution problems with the current solution that redis provides. The main example is the execution order and specially if you want transaction support. So lets say that you support transactions in a pipeline, then you must decide if you should support a full transaction across all slots and nodes in the cluster and that you have the execution order the very same as in a non cluster single server or not. So the main problem is that if you want to talk to 3 nodes with 3 different slots, then we right now split the command stack into each node and then executes each node at the same time. The problem is that if one node fails for any reason, then the other 2 nodes will execute and commit their transactions. So you are then left with a pipeline + transaction that 2/3 went through. So you can then do a second pass to correct the problem if the case was that a slot have moved to a new node for example. Yes you would have all the commands done in a transaction fashion, but you would not have the same execution order as you specificed in your code, if that is an issue for you. You could argue that this should not be a problem as you would have to have some crossslot commands to be executed for this to be relevant really.

I think i talk about these kinds of things in the docs for anyone to read up on, pipelines are super complex to get right if you want to provide a binary copy of the non clustered version of it. We made a choice about it when we developed it what it would be and the consequences of it. So people that uses this feature should be aware of the problems etc about it, but it is really hard to converse these very small and details about it really.

**Comment by Grokzen** [comment]
Closing it as an issue per say, but you can still ask if you have more questions about the topic.

**Comment by lodo1995** [comment]
@Grokzen thank you for your in-depth explanation.

You answered with the bit of explanation that I was looking for, that is, that in the worst case scenario I'm going to have 1 call for each node in the cluster. This is the part I was concerned about, as I'm going to have hundreds of commands in a single pipeline. I can stand having my pipeline make a request to each of my 4 or 5 nodes. What I can't stand is having my pipeline perform hundreds of calls to each of them.

Transactional and ordering guarantees don't concern my use-case, and in general I believe that fast in-memory databases like Redis should stay clear of this kind of complexity, and focus on providing the fastest solution to the simple key-value problem. Likewise, their users should design their system so that cached data does not need cross-machine guarantees. But that's another story.

Now I'm left with just another small question: if my pipeline needs to split the commands among 3 nodes, and thus perform 3 network calls, are these calls performed sequentially or in parallel?

**Comment by Grokzen** [comment]
Basically the only case where you will have this wierder case of pipeline logic is in the case you have a cluster node rebuild, or a node fails or you are manually moving a slot during a rebalance for example. If your cluster is stable then this should happen say less then 0.1% of the times or even less depending on the stability of your cluster. The logic that is implemented basically guarantees in a way that a pipeline will still function even in this kind of case. But there is never a 100% guarantee either as the pipeline will only retry so many times before it hard fails and at that point you get into the wierd scenarios as i described above.

The first implementation did a thread based style where a threadpool was used to run the commands in parallell, but this was later rebuilt to now do a kind of mixed style. What it will do is to write out the sockets in sequence, but write all 3 nodes at once before it starts to wait/listen/process the results. Yes it is still in hard sequential mode becuase the overhead of a pool or thread or any other solution would cost more then this solution that is still blazing fast compared to having to send off to one node, wait for it until sending off to the next node. I do not have any real numbers to back it up really, but in theory it should be as close to performant to using one node.

It is very true that not all cluster systems will work the same way across the board. CAP theory talks about that basically, and redis tallies on some of them to make something that works good in certain scnscenarios. But those tradeoffs is okey in a cache solution, but you would not use a redis cluster for true persistent data like your would have a SQL backend to back it up as a hard copy of it. Each one have to be designed and thought through about the benefits and limitations of the cluster solution.
